src/QsCompiler/CompilationManager/EditorSupport/SymbolInfo.cs

```diff
@@ -124,58 +124,70 @@ bool RangeOverlaps(QsNullable<Range> range)
-                QsStatementKind.QsReturnStatement @return => ExpressionDeclarations(@return.Item),
-                QsStatementKind.QsFailStatement fail => ExpressionDeclarations(fail.Item),
-                QsStatementKind.QsVariableDeclaration declaration => ExpressionDeclarations(declaration.Item.Rhs),
-                QsStatementKind.QsValueUpdate update => ExpressionDeclarations(update.Item.Rhs),
-                QsStatementKind.QsConditionalStatement cond => CondExpressionDeclarations(cond),
-                QsStatementKind.QsForStatement @for => ExpressionDeclarations(@for.Item.IterationValues),
-                QsStatementKind.QsWhileStatement @while => ExpressionDeclarations(@while.Item.Condition),
-                QsStatementKind.QsRepeatStatement repeat => ExpressionDeclarations(repeat.Item.SuccessCondition),
-                QsStatementKind.QsQubitScope qubit => QubitInitDeclarations(qubit.Item.Binding.Rhs),
-                _ => Enumerable.Empty<LocalVariableDeclaration<string>>(),
-            };
+                QsStatementKind.QsExpressionStatement e => ExpressionVars(e.Item),
+                QsStatementKind.QsReturnStatement r => ExpressionVars(r.Item),
+                QsStatementKind.QsFailStatement f => ExpressionVars(f.Item),
+                QsStatementKind.QsVariableDeclaration d => ExpressionVars(d.Item.Rhs),
+                QsStatementKind.QsValueUpdate u => ExpressionVars(u.Item.Rhs),
+                QsStatementKind.QsConditionalStatement c => CondExpressionVars(c),
+                QsStatementKind.QsForStatement f => ExpressionVars(f.Item.IterationValues),
+                QsStatementKind.QsWhileStatement w => ExpressionVars(w.Item.Condition),
+                QsStatementKind.QsRepeatStatement r => ExpressionVars(r.Item.SuccessCondition),
+                QsStatementKind.QsQubitScope q => QubitInitVars(q.Item.Binding.Rhs),
+                _ => null,
+            })?.Select(AddStatementOffset).Concat(varsBefore);
 
-            return new LocalDeclarations(
-                declarationsBefore
-                    .Concat(currentStatement.SymbolDeclarations.Variables)
-                    .Concat(expressionDeclarations.Select(AddStatementOffset))
-                    .ToImmutableArray());
+            var statementVars = currentStatement.Statement switch
+            {
+                QsStatementKind.QsForStatement f => f.Item.Body.KnownSymbols.Variables,
+                QsStatementKind.QsQubitScope q => q.Item.Body.KnownSymbols.Variables,
+                _ => currentStatement.SymbolDeclarations.Variables.Concat(varsBefore),
+            };
 
-            IEnumerable<LocalVariableDeclaration<string>> ExpressionDeclarations(TypedExpression expr) =>
-                DeclarationsInExpression(expr, position - currentStatement.Location.Item.Offset);
+            return new LocalDeclarations((expressionVars ?? statementVars).ToImmutableArray());
```

Comment:
This is the key part that fixes the bug - the previous version of LocalsInScope didn't include the child scope's environment of QsForStatement and QsQubitScope. However, we only want that if the position *isn't* in the statement's expression, so expressionVars is now null in that case. @ScottCarda-MS

Reply:
I see, that makes sense. Thanks!

Reply:
I see, that makes sense. Thanks!